### ISSUE TITLE
Net::Curl::Multi support (via Net::Curl::Promiser)

### ISSUE THREAD

**Issue by FGasper** [chazmcgarvey/HTTP-AnyUA#2]
Hello,

I’d like to add Net::Curl::Multi support via Net::Curl::Promiser. Would you accept a PR to this end?

**Comment by chazmcgarvey** [comment]
> I’d like to add Net::Curl::Multi support via Net::Curl::Promiser. Would you accept a PR to this end?

Yep, that would be great.

**Comment by FGasper** [comment]
@chazmcgarvey How does HTTP::AnyUA accept constructor arguments that HTTP::Tiny receives, like `max_redirect`?

**Comment by FGasper** [comment]
Or, more specifically, how does HTTP::AnyUA pass such parameters to the backend?

**Comment by chazmcgarvey** [comment]
HTTP::AnyUA doesn't actually construct any underlying user agents. It expects to receive an already-constructed user agent. Configuring the user agent is left to the user to do.

The backend just gets the already-configured user agent using https://metacpan.org/pod/HTTP::AnyUA::Backend#ua .

**Comment by FGasper** [comment]
@chazmcgarvey So how would an application that uses HTTP::AnyUA set `max_redirect=0` in a way that doesn’t depend on a particular HTTP client library? It doesn’t appear that there’s way to do this currently?

**Comment by chazmcgarvey** [comment]
Yeah, applications won't be user agent agnostic. They still need to choose which user agent to use. I'm not sure if there is value in an *application* being user agent agnostic... but I'm interested in what your use case is.

What HTTP::AnyUA can do is make *web service modules* work with any user agent the application creator may want to use. For example, if I create a web service module for the GitHub API I need a user agent but I don't want to force applications to use any particular user agent. Therefore my module would be useful in apps that want to use HTTP::Tiny and also in Mojolicious apps, etc.

Edit: See https://metacpan.org/pod/HTTP::AnyUA#How-do-I-set-up-proxying,-SSL,-cookies,-timeout,-etc.%3F more on this rationale.

**Comment by FGasper** [comment]
Sorry, by “application” I just meant “calling code”. :)

Some of what HTTP::Tiny expects as constructor arguments seem legitimately understood as _request_ properties. For example, `max_redirect`; I can discern no particularly good reason why HTTP::Tiny puts that parameter in the constructor without allowing individual requests to specify it. Really, `max_redirect` is a property of an individual request: one request might be fine to redirect 5 times; the next might want not to redirect at all.

The particular use case I’m looking at is teaching Net::ACME2 to work with parallel HTTP requests. Right now it uses HTTP::Tiny, so in principle your library would be a perfect fit to swap in.

`max_redirect` is a problem, though, because while Net::ACME2 may not need to care what actually sends the request per se, it needs whatever HTTP logic is in place to handle HTTP redirection limits according to whatever ACME server policy is in play. So if Net::ACME2 isn’t itself going to instantiate the internal UA (because its caller will determine whether HTTP::Tiny, Net::Curl::Promiser, or what not), then Net::ACME2 at least needs to be assured that redirections will be handled correctly.

This would seem a relatively straightforward matter to solve:
- Expose methods on HTTP::AnyUA that store such items as HTTP::Tiny’s `max_redirect` and `max_size`. (I suggest `set_configuration( $name, $value )`.)
- Expose that configuration to backends so that backends can implement the relevant functionality.

That’s probably a different feature request, of course.

**Comment by chazmcgarvey** [comment]
I see. Yeah, that's an interesting case.

Yeah, it could work if there were a generic interface each backend adheres to for configuring options and a pass-through method in HTTP::AnyUA.
